### DISCUSSION TITLE
Generic S-mode RISC-V EDK2 payload?

### DISCUSSION THREAD

**Discussion by LekKit** [tianocore/edk2#5815] [Ideas]
I see there is working support for RiscVVirt platform, which needs QEMU Virt + pflash device to run properly. However, there are other virtual machines without `cfi_flash` device emulation yet. Additionally, no actual RISC-V hardware has a `cfi_flash` device AFAIK.

Perhaps it makes sense to add another RiscVGeneric target which implements a generic S-mode payload to `fw_jump` OpenSBI stage, like U-Boot does for example?
I see there was something like that implemented fo FU540 board, which could be also booted on QEMU simply by passing a different `-bios` firmware: https://github.com/riscv-admin/riscv-uefi-edk2-docs?tab=readme-ov-file#edk2-risc-v-firmware-image-on-qemu.

Having a working upstread generic EDK2 S-mode payload would definitely help bring up support for other emulators like RVVM, and new actual hardware pieces.

**Comment by vlsunil** [comment]
qemu virt is the only platform supported in edk2 repo  mainly to aid the CI testing including NV variables. Otherwise, edk2-platforms should be used for any new reference/real platforms. U540 is not a good example for your requirement since it has the design where EDK2 starts in M-mode directly instead of as a payload . That design is not flexible and hence any new platform should use payload approach itself. There are  platforms already in edk2-platforms (ex: sg2042) which don't need flash like qemu virt but using payload design. They can be used as reference as well.

U540 need major cleanup and move to payload design. But currently, it is of low priority due to other tasks.

**Reply by LekKit** [reply]
Ah, thanks, it all makes much more sense now. I'll look at SG2042 platform soon, try to adapt it to my usecase and perhaps contribute a generic platform based on it if possible.

**Reply by vlsunil** [reply]
That will be great!. Thank you!
